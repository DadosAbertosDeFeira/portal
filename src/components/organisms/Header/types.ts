export type HeaderMenuItem = { href: string; label: string };
