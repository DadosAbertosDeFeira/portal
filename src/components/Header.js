import React, { useRef, useEffect, useState } from 'react';
import { BiMenu, BiSearchAlt2 } from 'react-icons/bi';
import { MdBusiness, MdClose } from 'react-icons/md';
import { animated, useTransition } from 'react-spring';

import styles from '../styles/components/header.module.scss';

export default function Header() {
  const navRef = useRef(null);
  const imgRef = useRef(null);

  const [menuOpened, setMenuOpened] = useState(false);
  const [searchOpened, setSearchOpened] = useState(false);

  const [searchText, setSearchText] = useState('');

  const toggleMenu = () => {
    setMenuOpened((state) => !state);
  };

  const openSearch = () => {
    setSearchOpened(true);
  };
  const closeSearch = () => {
    setSearchOpened(false);
  };

  const menuTransition = useTransition(menuOpened, null, {
    from: {
      left: '-500px',
    },
    enter: {
      left: '0',
    },
    leave: {
      left: '-500px',
    },
  });

  const searchTransition = useTransition(!searchOpened, null, {
    from: {
      transform: 'translateX(120%)',
    },
    enter: {
      position: 'static',
      transform: 'translateX(0)',
    },
    leave: {
      position: 'absolute',
      transform: 'translateX(-150%)',
    },
  });

  useEffect(() => {
    let isScrollingDown = false;

    const onScroll = () => {
      const scrollTop = window.pageYOffset || navRef.scrollTop;
      if (scrollTop > 68) {
        if (isScrollingDown) {
          return;
        }
        isScrollingDown = true;
        navRef.current.classList.add(styles.decreasedSize);
        navRef.current.classList.remove(styles.defaultSize);
        closeSearch();
      } else {
        if (!isScrollingDown) {
          return;
        }
        isScrollingDown = false;
        navRef.current.classList.add(styles.defaultSize);
        navRef.current.classList.remove(styles.decreasedSize);
        closeSearch();
      }
    };
    window.addEventListener('scroll', onScroll);
    return () => {
      window.removeEventListener('scroll', onScroll);
    };
  }, [navRef]);

  return (
    <>
      <nav id={styles.navbar} className={styles.defaultSize} ref={navRef}>
        <header>
          {searchTransition.map(
            ({ item, key, props }) =>
              !item && (
                <animated.div key={key} style={props}>
                  <div>
                    <img
                      className={styles.decreasedSize}
                      src="assets/logo_pequena.svg"
                      alt="Logo pequena Dados Abertos de Feira"
                    />
                  </div>
                  <form action="busca">
                    <BiSearchAlt2 size={24} />
                    <input
                      ref={(input) => input && input.focus()}
                      name="search"
                      placeholder="Digite aqui a sua busca"
                      onChange={(e) => setSearchText(e.target.value)}
                      value={searchText}
                    />
                    <button type="button" onClick={closeSearch}>
                      <MdClose size={24} />
                    </button>
                  </form>
                </animated.div>
              )
          )}
          {searchTransition.map(
            ({ item, key, props }) =>
              item && (
                <animated.div key={key} style={props}>
                  <button type="button" onClick={toggleMenu}>
                    <BiMenu size={24} />
                  </button>
                  <div id={styles.reducedImageWithScroll}>
                    <img
                      src="assets/logo.svg"
                      alt="Logo Dados Abertos de Feira"
                      ref={imgRef}
                    />
                  </div>
                  <button type="button" onClick={openSearch}>
                    <BiSearchAlt2 size={24} />
                  </button>
                </animated.div>
              )
          )}
        </header>
      </nav>
      {menuTransition.map(
        ({ item, key, props }) =>
          item && (
            <animated.div id={styles.menu} key={key} style={props}>
              <div id={styles.menuClose}>
                <button type="button" onClick={toggleMenu}>
                  <MdClose size={24} />
                </button>
              </div>
              <ul className={styles.list}>
                <li>
                  <div>
                    <MdBusiness size={16} />
                    <span>Prefeitura</span>
                  </div>
                  <ul className={styles.listItem}>
                    <li>
                      <span>Diário Oficial</span>
                    </li>
                    <li>
                      <span>Licitações</span>
                    </li>
                  </ul>
                </li>

                <li>
                  <div>
                    <MdBusiness size={16} />
                    <span>Câmara de Vereadores</span>
                  </div>
                  <ul className={styles.listItem}>
                    <li>
                      <span>Diário Oficial</span>
                    </li>
                    <li>
                      <span>Atas das Sessões</span>
                    </li>
                    <li>
                      <span>Lista de Presença</span>
                    </li>
                    <li>
                      <span>Agenda das Sessões</span>
                    </li>
                  </ul>
                </li>
              </ul>
            </animated.div>
          )
      )}
    </>
  );
}
