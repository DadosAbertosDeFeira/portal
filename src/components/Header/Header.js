import React, { useState } from 'react';
import { BiMenu } from 'react-icons/bi';
import Link from 'next/link';
import MobileMenu from '../MobileMenu';

const DesktopHeader = () => {
  return (
    <nav className="bg-white fixed h-desktop-header top-0 z-20 xl:z-0 w-full flex-row shadow md:space-x-lg page-margin-align pr-xl hidden tablet:flex">
      <div className="my-auto mr-lg">
        <Link href="/">
          <img
            className="h-14 w-auto"
            src="/assets/logo.svg"
            alt="Logo Dados Abertos de Feira"
          />
        </Link>
      </div>
      <ul className="flex flex-row justify-start space-x-xl my-auto font-bold text-sm z-20">
        <li>
          <Link href="/#sobre">Sobre</Link>
        </li>
        <li>
          <Link href="/#voluntarios">Volunt√°rios</Link>
        </li>
        <li>
          <Link href="/#projetos">Projetos</Link>
        </li>
        <li>
          <Link href="/#colabore">Colabore</Link>
        </li>
        <li>
          <Link href="/#contato">Contato</Link>
        </li>
        <li>
          <Link target="_blank" href="https://anchor.fm/cidadesabertas/">
            Podcast
          </Link>
        </li>
        <li>
          <Link target="_blank" href="https://dadosabertosdefeira.medium.com">
            Blog
          </Link>
        </li>
      </ul>
    </nav>
  );
};

const MobileHeader = () => {
  const [menuOpened, setMenuOpened] = useState(false);

  const toggleMenu = () => {
    setMenuOpened((state) => !state);
  };

  const closeMenuIfOpened = () => {
    setMenuOpened((isOpen) => (isOpen ? false : isOpen));
  };

  return (
    <nav className="bg-white fixed top-0 h-mobile-header w-full z-10 flex-row px-lg shadow flex tablet:hidden">
      <div className="flex-grow flex place-content-center">
        <Link href="/" className="m-auto">
          <img
            src="/assets/logo.svg"
            className="h-14 w-auto"
            alt="Logo Dados Abertos de Feira"
          />
        </Link>
      </div>
      <button type="button" onClick={toggleMenu}>
        <BiMenu size={24} />
      </button>
      <aside
        className={`fixed top-0 left-0 z-20 h-full ease-in-out transition-all duration-300
          ${
            menuOpened ? 'visible translate-x-0' : 'invisible -translate-x-full'
          }`}
      >
        <MobileMenu handleClose={closeMenuIfOpened} />
      </aside>
    </nav>
  );
};

/**
 * Componente do Header
 *
 * @param props {}
 *
 * @example
 * ```
 * import Header from "../components/Header";
 *
 * export default function Page() {
 *
 *   return (
 *     <>
 *       ...
 *       <Header />
 *       ...
 *     </>
 *   )
 * }
 * ```
 */

export default function Header() {
  return (
    <header>
      <DesktopHeader />
      <MobileHeader />
    </header>
  );
}
