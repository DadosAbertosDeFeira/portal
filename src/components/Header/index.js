import React, { useRef, useEffect, useState } from 'react';
import { BiMenu } from 'react-icons/bi';
import { FaFacebook, FaInstagram, FaTwitter } from 'react-icons/fa';
import { animated, useTransition } from 'react-spring';
import Link from '../Link';

import styles from './styles.module.scss';
import SearchDropdown from '../SearchDropdown';

import logo from '../../assets/logo.svg';
import smallLogo from '../../assets/smallLogo.svg';

/**
 * Componente do Dropdown de Pesquisa
 *
 * @param props {}
 *
 * @example
 * ```
 * import Header from "../components/Header";
 *
 * export default function Page() {
 *
 *   return (
 *     <>
 *       ...
 *       <Header />
 *       ...
 *     </>
 *   )
 * }
 * ```
 */

export default function Header() {
  const navRef = useRef(null);
  const imgRef = useRef(null);

  const [menuOpened, setMenuOpened] = useState(false);
  const [searchOpened, setSearchOpened] = useState(false);

  // const [searchText, setSearchText] = useState('');

  const toggleMenu = () => {
    setMenuOpened((state) => !state);
  };

  // const openSearch = () => {
  //   setSearchOpened(true);
  // };

  const closeSearch = () => {
    setSearchOpened(false);
  };

  const searchTransition = useTransition(!searchOpened, null, {
    from: {
      transform: 'translateX(150%)',
    },
    enter: {
      position: 'static',
      transform: 'translateX(0)',
    },
    leave: {
      position: 'absolute',
      transform: 'translateX(-150%)',
    },
  });

  useEffect(() => {
    let isScrollingDown = false;

    const onScroll = () => {
      if (!navRef.current) {
        return;
      }
      const scrollTop = window.pageYOffset || navRef.scrollTop;
      if (scrollTop > 68) {
        if (isScrollingDown) {
          return;
        }
        isScrollingDown = true;
        navRef.current.classList.add(styles.decreasedSize);
        navRef.current.classList.remove(styles.defaultSize);
        closeSearch();
      } else {
        if (!isScrollingDown) {
          return;
        }
        isScrollingDown = false;
        navRef.current.classList.add(styles.defaultSize);
        navRef.current.classList.remove(styles.decreasedSize);
        closeSearch();
      }
    };
    window.addEventListener('scroll', onScroll);

    return () => {
      window.removeEventListener('scroll', onScroll);
    };
  }, [navRef]);

  return (
    <>
      <header id={styles.header} className={styles.defaultSize} ref={navRef}>
        <nav id={styles.desktop}>
          <div>
            <Link href="/">
              <img
                className={styles.logo}
                src={logo}
                alt="Logo Dados Abertos de Feira"
              />
            </Link>
          </div>
          <ul>
            <li>
              <Link href="/#sobre">SOBRE</Link>
            </li>
            <li>
              <Link
                target="_blank"
                href="https://dadosabertosdefeira.medium.com"
              >
                BLOG
              </Link>
            </li>
            <li>
              <Link href="/#colabore">COLABORE</Link>
            </li>
          </ul>
          <ul className={styles.social}>
            <li>
              <Link href={process.env.NEXT_PUBLIC_INSTAGRAM_URL}>
                <FaInstagram size={32} />
              </Link>
            </li>
            <li>
              <Link href={process.env.NEXT_PUBLIC_FACEBOOK_URL}>
                <FaFacebook size={32} />
              </Link>
            </li>
            <li>
              <Link href={process.env.NEXT_PUBLIC_TWITTER_URL}>
                <FaTwitter size={32} />
              </Link>
            </li>
          </ul>
          <form action="busca">
            {/* <BiSearchAlt2 size={32} />
            <input
              ref={(input) => input && input.focus()}
              name="search"
              placeholder="Pesquisar no Dados Abertos de Feira"
              onChange={(e) => setSearchText(e.target.value)}
              value={searchText}
            /> */}
          </form>
        </nav>
        <nav id={styles.mobile}>
          {searchTransition.map(
            ({ item, key, props }) =>
              !item && (
                <animated.div key={key} style={props}>
                  <div>
                    <Link href="/">
                      <img
                        className={styles.searchImageSize}
                        src={smallLogo}
                        alt="Logo pequena Dados Abertos de Feira"
                      />
                    </Link>
                  </div>
                  <form action="busca">
                    {/* <BiSearchAlt2 size={24} />
                    <input
                      ref={(input) => input && input.focus()}
                      name="search"
                      placeholder="Digite aqui a sua busca"
                      onChange={(e) => setSearchText(e.target.value)}
                      value={searchText}
                    />
                    <button type="button" onClick={closeSearch}>
                      <MdClose size={24} />
                    </button> */}
                  </form>
                </animated.div>
              )
          )}
          {searchTransition.map(
            ({ item, key, props }) =>
              item && (
                <animated.div key={key} style={props}>
                  {/* <button type="button" onClick={openSearch}>
                    <BiSearchAlt2 size={24} />
                  </button> */}
                  <div />
                  <div id={styles.reducedImageWithScroll}>
                    <Link href="/">
                      <img
                        src={logo}
                        alt="Logo Dados Abertos de Feira"
                        ref={imgRef}
                      />
                    </Link>
                  </div>
                  <button type="button" onClick={toggleMenu}>
                    <BiMenu size={24} />
                  </button>
                </animated.div>
              )
          )}
        </nav>
      </header>
      <div id={styles.menu} className={menuOpened ? styles.open : ''}>
        <SearchDropdown toggleMenu={toggleMenu} />
      </div>
    </>
  );
}
