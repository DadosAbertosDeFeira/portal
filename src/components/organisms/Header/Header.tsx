import type { BoxProps } from "atoms/Box";
import { Box } from "atoms/Box";
import { Logo } from "atoms/Logo";
import { Drawer } from "molecules/Drawer";
import { SandwichMenu } from "molecules/SandwichMenu";
import React, { useEffect, useMemo, useRef, useState } from "react";
import { createPortal } from "react-dom";

import { useMediaQuery } from "@/hooks/useMediaQuery";
import { useOnClickOutside } from "@/hooks/useOnClickOutside";

import { HeaderAnchors } from "./HeaderAnchors";
import { HeaderMenu } from "./HeaderMenu";

export type HeaderProps = BoxProps & {};

export function Header(props: HeaderProps) {
  const drawerRef = useRef<HTMLDivElement | null>(null);
  const sandwichRef = useRef<HTMLButtonElement | null>(null);
  const root = useRef<HTMLElement | null>(null);

  useEffect(() => {
    root.current = document.body;
  }, [root]);

  const [isOpen, setOpen] = useState(false);
  const matches = useMediaQuery("(min-width: 768px)");

  useOnClickOutside(drawerRef, (event) => {
    const target = event.target as HTMLElement;
    const sandwich = sandwichRef.current;
    if (sandwich === null) return;

    if (!sandwich.contains(target)) setOpen(false);
  });

  const headerContent = useMemo(() => {
    if (matches) return <HeaderAnchors />;
    return (
      <SandwichMenu
        ref={sandwichRef}
        isOpen={isOpen}
        onClick={() => setOpen((oldState) => !oldState)}
      />
    );
  }, [matches, isOpen]);

  return (
    <Box
      className="fixed inset-x-0 top-0 z-20 flex h-[10vh] max-h-[80px] min-h-[60px] flex-row gap-3 bg-white px-6 shadow"
      {...props}
    >
      <Box className="flex w-[150px] grow items-center justify-center md:mr-10 md:w-[15vw] md:grow-0 md:justify-end">
        <Logo className="w-[150px]" />
      </Box>
      {headerContent}
      {root.current &&
        createPortal(
          <Drawer ref={drawerRef} isOpen={isOpen}>
            <HeaderMenu closeMenu={() => setOpen(false)} />
          </Drawer>,
          root.current
        )}
    </Box>
  );
}
