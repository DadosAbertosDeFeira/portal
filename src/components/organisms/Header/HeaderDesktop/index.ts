export * from "./HeaderDesktop";
export * from "./types";
