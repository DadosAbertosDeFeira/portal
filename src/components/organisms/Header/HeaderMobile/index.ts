export * from "./HeaderMobile";
export * from "./types";
