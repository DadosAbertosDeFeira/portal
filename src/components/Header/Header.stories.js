import React from "react";
import {
  INITIAL_VIEWPORTS,
  MINIMAL_VIEWPORTS,
} from "@storybook/addon-viewport";

import Header from ".";

export default {
  title: "Components/Header",
  component: Header,
  parameters: {
    viewport: {
      viewports: {
        ...INITIAL_VIEWPORTS,
        ...MINIMAL_VIEWPORTS,
      },
      defaultViewport: "responsive",
    },
    docs: {
      inlineStories: false,
      iframeHeight: 500,
    },
  },
};

export const Default = () => <Header />;
