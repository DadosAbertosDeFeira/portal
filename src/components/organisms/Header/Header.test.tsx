import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Header, type HeaderProps } from "./Header";

describe("Header", () => {
  let user = userEvent.setup();

  beforeEach(() => {
    user = userEvent.setup();
    jest.clearAllMocks();
  });

  const makeSut = ({ ...props }: Partial<HeaderProps>) => {
    return render(<Header {...props} />);
  };

  it("renders correctly", () => {
    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });

  it("should render desktop menu correctly", () => {
    window.matchMedia = jest.fn().mockImplementation(() => ({
      matches: true,
      media: "",
      onchange: null,
      addEventListener: jest.fn((_, handler) => handler({ matches: true })),
      removeEventListener: jest.fn(),
    }));

    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });

  it("should drawer have left-0 when click in button", async () => {
    makeSut({});

    const button = screen.getByRole("button", { name: "Menu" });

    await act(async () => user.click(button));

    const drawer = screen.getByRole("navigation");

    expect(drawer).toHaveClass("left-0");
  });

  it("should close drawer menu when click outside", async () => {
    const { container } = makeSut({});

    const button = screen.getByRole("button", { name: "Menu" });

    await act(async () => user.click(button));
    await act(async () => user.click(container));

    const drawer = screen.getByRole("navigation");

    expect(drawer).toHaveClass("left-[-70vw]");
  });
});
