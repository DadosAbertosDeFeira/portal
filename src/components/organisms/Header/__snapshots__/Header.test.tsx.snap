// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `[Function]`;

exports[`Header should render desktop menu correctly 1`] = `[Function]`;
